# Data Analysis with Pandas

## Project Description
In this project, we will work with data from the entertainment industry. We will study a dataset with records on movies and shows. The research will focus on the “Golden Age” of television, which began in 1999 with the release of The Sopranos and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives from IMDb users impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the “Golden Age” of television also have the most votes.

Many business decisions are initially born as assumptions, our contributions as experts in the data domain is to answer the question “Did the assumption formulated before the study appear to be true?”

## Data Description
The dataset movies_and_shows.csv contains information about various movies and shows, including:

- name: The name of the actor or actress.
- Character: The character they played.
- r0le: The role type (e.g., ACTOR).
- TITLE: The title of the movie or show.
- Type: Whether it's a MOVIE or SHOW.
- release Year: The year it was released.
- genres: A list of genres the movie or show belongs to.
- imdb sc0re: The IMDb score of the movie or show.
- imdb v0tes: The number of votes on IMDb.

# Imports & Setup

In [2]:
# Importing pandas
import pandas as pd

In [3]:
# Reading .csv File
df = pd.read_csv('movies_and_shows.csv')

# Data Overview

In [4]:
# Breifly looking at the data
df.head()

,name,Character,r0le,TITLE,Type,release Year,genres,imdb sc0re,imdb v0tes
0,Robert De Niro,Travis Bickle,ACTOR,Taxi Driver,MOVIE,1976,"['drama', 'crime']",8.2,808582.0
1,Jodie Foster,Iris Steensma,ACTOR,Taxi Driver,MOVIE,1976,"['drama', 'crime']",8.2,808582.0
2,Albert Brooks,Tom,ACTOR,Taxi Driver,MOVIE,1976,"['drama', 'crime']",8.2,808582.0
3,Harvey Keitel,Matthew 'Sport' Higgins,ACTOR,Taxi Driver,MOVIE,1976,"['drama', 'crime']",8.2,808582.0
4,Cybill Shepherd,Betsy,ACTOR,Taxi Driver,MOVIE,1976,"['drama', 'crime']",8.2,808582.0


In [5]:
# Checking the number of rows and columns
df.shape

(85579, 9)

In [6]:
# Looking at the datas information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


After looking at the data:
- There are 85579 rows and 9 columns, with each row representing an actor/actress. 
- At a glance, we can see:
    - Column names need to be handled.
    - There are missing values for 'imdb score' & 'imdb votes'.

# Data Cleaning


## Fixing Column Names

In [9]:
# Let's get the column names
df.columns

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

In [10]:
# Rename columns to make them consistent.
df = df.rename(columns = {
    '   name': 'name',
    'Character': 'character',
    'r0le': 'role',
    'TITLE': 'title',
    '  Type': 'type',
    'release Year': 'release_year',
    'genres': 'genre',
    'imdb sc0re': 'imdb_score',
    'imdb v0tes': 'imdb_votes'
})

# Looking at the updated column names
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genre',
       'imdb_score', 'imdb_votes'],
      dtype='object')

## Duplicates?

In [11]:
# Looking to see if there are any duplicate rows
df.duplicated().sum()

np.int64(7675)

In [15]:
# Dropping the duplicate rows
df = df.drop_duplicates()
# Looking to see if there are any duplicate rows
df.duplicated().sum()

np.int64(0)

## Missing Values?

In [16]:
# Looking to see if there are missing values
df.isnull().sum()

name               0
character          0
role               0
title              1
type               0
release_year       0
genre              0
imdb_score      3958
imdb_votes      4045
dtype: int64

In [21]:
# Let's look at the rows with missing values
# 'title' column
print("Missing values in 'title' column:", df[df['title'].isnull() == True])

# 'imdb_score' column
print("Missing values in 'imdb_score' column:", df[df['imdb_score'].isnull() == True])

# 'imdb_votes' column
print("Missing values in 'imdb_votes' column:", df[df['imdb_votes'].isnull() == True])

Missing values in 'title' column:                name character      role title   type  release_year genre  \
27556  J??rn Donner   unknown  DIRECTOR   NaN  MOVIE          2015    []   

       imdb_score  imdb_votes  
27556         NaN         NaN  
Missing values in 'imdb_score' column:                      name               character      role  \
1478           Bill Hicks                 Himself     ACTOR   
1479          Kevin Booth                 unknown  DIRECTOR   
1480       David Johndrow                 unknown  DIRECTOR   
4052    Alejandra Borrero                Consuelo     ACTOR   
4053   Juan Carlos Arango  Jaime "El Mane" Molina     ACTOR   
...                   ...                     ...       ...   
84658        Sohayb Godus                 unknown     ACTOR   
84659     Shahad Alahmari                 unknown     ACTOR   
84660         Malik Nejer                 unknown  DIRECTOR   
84689       Baek Jong-won                    Self     ACTOR   
84690       Park 

In [22]:
# We can replace the missing value of 'title' with 'unknown'
df['title'] = df['title'].fillna('unknown')

# Since 'imdb_score' and 'imdb_votes' are missing, we will drop the rows with missing values.
df = df.dropna(subset=['imdb_score', 'imdb_votes'])

# Looking to see if there are any missing values
df.isnull().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genre           0
imdb_score      0
imdb_votes      0
dtype: int64

In [24]:
# Checking the number of rows and columns after cleaning
print("Number of rows and columns after cleaning:", df.shape)
# Data Analysis
# Let's look at the data
df.head()

Number of rows and columns after cleaning: (73859, 9)


,name,character,role,title,type,release_year,genre,imdb_score,imdb_votes
0,Robert De Niro,Travis Bickle,ACTOR,Taxi Driver,MOVIE,1976,"['drama', 'crime']",8.2,808582.0
1,Jodie Foster,Iris Steensma,ACTOR,Taxi Driver,MOVIE,1976,"['drama', 'crime']",8.2,808582.0
2,Albert Brooks,Tom,ACTOR,Taxi Driver,MOVIE,1976,"['drama', 'crime']",8.2,808582.0
3,Harvey Keitel,Matthew 'Sport' Higgins,ACTOR,Taxi Driver,MOVIE,1976,"['drama', 'crime']",8.2,808582.0
4,Cybill Shepherd,Betsy,ACTOR,Taxi Driver,MOVIE,1976,"['drama', 'crime']",8.2,808582.0


# Basic Exploration
To get familiar with working with data, we will only be doing the following: 
- Find Highly Rates Movies
- Find Unique Top-Rated Movies
- Find Top Movies from a Specific Decade
- List all Actors in a Given Title
- Categorize Movies by IMDb Score

## Finding Highly Rated Movies
We want to identify movies with an IMDb rating of at least 9.0. This list could be helpful for curating a 'Top Movies' section based on high ratings.

In [29]:
#Filtering for movies with IMDb score above 9.0
high_score_df = df[df['imdb_score'] >= 9.0]

#Extract the 'title' column
high_score_titles = df.loc[df['imdb_score'] >= 9.0, 'title']

# Get a unique set of titles
unique_titles = set(high_score_titles)

# Print the unique titles
print(unique_titles)

{'DEATH NOTE', 'Leah Remini: Scientology and the Aftermath', 'Kota Factory', 'Reply 1988', 'My Mister', 'Major', 'Avatar: The Last Airbender', 'Our Planet', 'Arcane', 'The Last Dance', 'Attack on Titan', 'Hunter x Hunter', 'Breaking Bad', 'Okupas'}


## Find Unique Top-Rated Movies
We'll create a function to find unique movies with an IMDb score above a certain threshold that the user provides. 

In [35]:
def get_unique_top_movies(min_score):
    # Filter for movies with IMDb score above min_score
    high_score_df = df[df['imdb_score'] >= min_score]

    # Extract the 'title' column
    high_score_titles = df.loc[df['imdb_score'] >= min_score, 'title']
    
    # Extract titles
    high_score_titles = set(high_score_titles)

    # Return unique titles
    return list(high_score_titles)

In [36]:
# Test the function
print(get_unique_top_movies(9.0)) # Should be the same as the list we got from the previous section.

['DEATH NOTE', 'Leah Remini: Scientology and the Aftermath', 'Kota Factory', 'Reply 1988', 'My Mister', 'Major', 'Avatar: The Last Airbender', 'Our Planet', 'Arcane', 'The Last Dance', 'Attack on Titan', 'Hunter x Hunter', 'Breaking Bad', 'Okupas']


## Find Top Movies from a Specific Decade
We'll create a function to retrieve movies from a particular decade with high IMDb ratings. This function can be useful for generating lits of top-rated movies from different time periods. 

In [37]:
def get_top_movies_from_decade(decade_start, min_score):
    # Filter for movies released within the decade
    decade = df[(df['release_year'] >= decade_start) & (df['release_year'] <= decade_start + 9)]

    # Further filter by IMDb score
    decade = decade[decade['imdb_score'] >= min_score]
    
    # Extract titles
    titles = set(decade['title'])

    # Return unique titles
    return list(titles)


In [38]:
# Test the function
print(get_top_movies_from_decade(1990, 8.5))

['One Piece', 'Cowboy Bebop', 'Bill Hicks: Revelations', 'Se7en', 'Neon Genesis Evangelion', 'GoodFellas', 'Forrest Gump', 'L??on: The Professional']


## List All Actors in a Given Title
We'll create a function that takes a title as input and returns the names of all actors in that title, combined into a single string. 

In [49]:
def get_actors_in_title(title):
    title_df = df[(df['title'] == title) & (df['role'] == 'ACTOR')]

    # Extract the 'name' column for actor names
    title_df = title_df['name']

    # Combine actor names into a single string
    actors = ', '.join(title_df.to_list())

    # Return the list of actors
    return actors

In [50]:
# Test the function
get_actors_in_title('Taxi Driver')

'Robert De Niro, Jodie Foster, Albert Brooks, Harvey Keitel, Cybill Shepherd, Peter Boyle, Leonard Harris, Diahnne Abbott, Gino Ardito, Martin Scorsese, Murray Moston, Richard Higgs, Bill Minkin, Bob Maroff, Victor Argo, Joe Spinell, Robinson Frank Adu, Brenda Dickson, Norman Matlock, Harry Northup, Harlan Cary Poe, Steven Prince, Peter Savage, Nicholas Shields, Ralph S. Singleton, Annie Gagen, Carson Grant, Mary-Pat Green, Debbi Morgan, Don Stroud, Copper Cunningham, Garth Avery, Nat Grant, Billie Perkins, Catherine Scorsese, Charles Scorsese, Odunlade Adekola, Ijeoma Grace Agu'

## Categorize Movies by IMDb Score
Let's categorize movies and shows based on their IMDb scores to provide a quick evaluation of their popularity or quality.

**Rating Categories**
- Excellent: IMDb score of 9.0 or higher
- Good: IMDb score between 7.0 and 8.9
- Average: IMDb score between 5.0 and 6.9
- Low: IMDb score below 5.0

In [71]:
def categorize_movies_by_score(title):
    # Filter for the specific title
    title_df = df[df['title'] == title]

    # Check if the title exists in the dataset
    if title_df.empty:
        return "Title not found in the dataset."

    # Build one result per row using that row's name and score
    results = []
    for _, row in title_df.iterrows():
        name = row['name']
        score = row['imdb_score']
        if score >= 9.0:
            results.append(f'{name} - Score: {score} - Excellent')
        elif score >= 7.0 and score < 9.0:
            results.append(f'{name} - Score: {score} - Good')
        elif score >= 5.0 and score < 7.0:
            results.append(f'{name} - Score: {score} - Average')
        else:
            results.append(f'{name} - Score: {score} - Low')

    return results
        

In [72]:
# Test the function
print(categorize_movies_by_score('Taxi Driver'))

['Robert De Niro - Score: 8.2 - Good', 'Jodie Foster - Score: 8.2 - Good', 'Albert Brooks - Score: 8.2 - Good', 'Harvey Keitel - Score: 8.2 - Good', 'Cybill Shepherd - Score: 8.2 - Good', 'Peter Boyle - Score: 8.2 - Good', 'Leonard Harris - Score: 8.2 - Good', 'Diahnne Abbott - Score: 8.2 - Good', 'Gino Ardito - Score: 8.2 - Good', 'Martin Scorsese - Score: 8.2 - Good', 'Murray Moston - Score: 8.2 - Good', 'Richard Higgs - Score: 8.2 - Good', 'Bill Minkin - Score: 8.2 - Good', 'Bob Maroff - Score: 8.2 - Good', 'Victor Argo - Score: 8.2 - Good', 'Joe Spinell - Score: 8.2 - Good', 'Robinson Frank Adu - Score: 8.2 - Good', 'Brenda Dickson - Score: 8.2 - Good', 'Norman Matlock - Score: 8.2 - Good', 'Harry Northup - Score: 8.2 - Good', 'Harlan Cary Poe - Score: 8.2 - Good', 'Steven Prince - Score: 8.2 - Good', 'Peter Savage - Score: 8.2 - Good', 'Nicholas Shields - Score: 8.2 - Good', 'Ralph S. Singleton - Score: 8.2 - Good', 'Annie Gagen - Score: 8.2 - Good', 'Carson Grant - Score: 8.2 - G